include/onnxruntime/core/session/onnxruntime_cxx_api.h

```diff
@@ -1051,6 +1054,54 @@ struct ArenaCfg : Base<OrtArenaCfg> {
 // Custom OPs (only needed to implement custom OPs)
 //
 
+/// <summary>
+/// This struct wraps a raw pointer const OrtNodeArg* that is obtained from
+/// a KernelInfo object. Use it to access the name or type info of an input
+/// or output.
+/// </summary>
+struct NodeArg {
+  explicit NodeArg(const OrtNodeArg* node_arg);
+
+  // Returns a pair consisting of a Ort::AllocatedStringPtr and the string's length.
+  // Ex: auto name = node_arg.GetName(allocator);
+  // name.first.get(); // Get char*
+  // name.second; // Get length
+  std::pair<AllocatedStringPtr, size_t> GetName(OrtAllocator* allocator) const;
+
+  // Returns the node's type and shape information.
+  [[nodiscard]] TypeInfo GetTypeInfo() const;
+
+  constexpr explicit operator const OrtNodeArg*() const noexcept { return p_; }
+
+ private:
+  const OrtNodeArg* p_;
+};
+
+/// <summary>
+/// This struct wraps a const OrtKernelInfo*.
+/// TODO: Use Base<T> to implement OrtKernelInfoCopy()
+/// </summary>
+struct KernelInfo {
+  explicit KernelInfo(const OrtKernelInfo* info);
+
+  template <typename T>  // T is only implemented for float, int64_t, and string
+  T GetAttribute(const char* name) const;
+
+  template <typename T>  // T is only implemented for std::vector<float>, std::vector<int64_t>
+  std::vector<T> GetAttributes(const char* name) const;
+
+  [[nodiscard]] size_t GetInputCount() const;
```

Comment:
>[[nodiscard] [](http://example.com/codeflow?start=2&length=12)

We disabled clang tidy. This would require users to have C++17.

Reply:
I'll remove [[discard]]